### PR TITLE
[8.0] fix: wait and retry on ConflictError

### LINKED ISSUE #6715: Conflicts when updating JobParameters in ElasticJobParametersDB
In LHCb we've seen the below callstack happening from time to time. I think it's a race between different RPC calls that asynchronously set the `Status` parameter. There is a `retry_on_conflict` option in elasticsearch but I'm not convinced this is the right thing to do.

```python
2023-01-24 07:39:07 UTC WorkloadManagement/JobStateUpdate ERROR: Uncaught exception when serving RPC Function setJobStatusBulk
Traceback (most recent call last):
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/Core/DISET/RequestHandler.py", line 296, in __RPCCallFunction
    uReturnValue = oMethod(*args)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/WorkloadManagementSystem/Service/JobStateUpdateHandler.py", line 137, in export_setJobStatusBulk
    return cls._setJobStatusBulk(jobID, statusDict, force=force)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/WorkloadManagementSystem/Service/JobStateUpdateHandler.py", line 257, in _setJobStatusBulk
    result = cls.elasticJobParametersDB.setJobParameter(int(jobID), "Status", status)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/WorkloadManagementSystem/DB/ElasticJobParametersDB.py", line 177, in setJobParameter
    result = self.updateDoc(index=self._indexName(jobID), docID=str(jobID), body={"doc": data})
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/Core/Utilities/ElasticSearchDB.py", line 44, in wrapper_decorator
    return method(self, *args, **kwargs)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/Core/Utilities/ElasticSearchDB.py", line 269, in updateDoc
    return S_OK(self.client.update(index, docID, body))
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/opensearchpy/client/utils.py", line 178, in _wrapped
    return func(*args, params=params, headers=headers, **kwargs)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/opensearchpy/client/__init__.py", line 1767, in update
    return self.transport.perform_request(
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/opensearchpy/transport.py", line 408, in perform_request
    raise e
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/opensearchpy/transport.py", line 369, in perform_request
    status, headers_response, data = connection.perform_request(
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/opensearchpy/connection/http_urllib3.py", line 266, in perform_request
    self._raise_error(
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/opensearchpy/connection/base.py", line 301, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(
opensearchpy.exceptions.ConflictError: ConflictError(409, 'version_conflict_engine_exception', '[709792449]: version conflict, required seqNo [16420917], primary term [1]. current document has seqNo [16421015] and primary term [1]')
2023-01-24 07:39:07 UTC WorkloadManagement/JobStateUpdate NOTICE: Returning response ([::ffff:81.180.86.124]:54022)[lhcb_mc:fstagni] (0.30 secs) ERROR: Server error while serving setJobStatusBulk: ConflictError(409, 'version_conflict_engine_exce
ption', '[709792449]: version conflict, required seqNo [16420917], primary term [1]. current document has seqNo [16421015] and primary term [1]')
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/threading.py", line 937, in _bootstrap
    self._bootstrap_inner()
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/threading.py", line 980, in _bootstrap_inner
    self.run()
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/threading.py", line 917, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/concurrent/futures/thread.py", line 83, in _worker
    work_item.run()
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/Core/DISET/private/Service.py", line 351, in _processInThread
    result = self._processProposal(trid, proposalTuple, handlerObj)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/Core/DISET/private/Service.py", line 538, in _processProposal
    result = self._executeAction(trid, proposalTuple, handlerObj)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/Core/DISET/private/Service.py", line 558, in _executeAction
    response = handlerObj._rh_executeAction(proposalTuple)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/Core/DISET/RequestHandler.py", line 124, in _rh_executeAction
    retVal = self.__doRPC(actionTuple[1])
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/Core/DISET/RequestHandler.py", line 255, in __doRPC
    return self.__RPCCallFunction(method, args)
  File "/opt/dirac/versions/v11.0.0-1674461755/Linux-x86_64/lib/python3.9/site-packages/DIRAC/Core/DISET/RequestHandler.py", line 306, in __RPCCallFunction
    return S_ERROR(f"Server error while serving {method}: {str(e)}")
```

Screenshot 2023-01-24 at 10 31 37

### PR DESCRIPTION
**Pull request by fstagni**
Proposal to fix https://github.com/DIRACGrid/DIRAC/issues/6715

BEGINRELEASENOTES

*Monitoring
FIX: while updating ES documents, wait and retry on ConflictError

ENDRELEASENOTES

closes #6715

### CODE REVIEW

**Comment by chaen** [comment]
Just a question, but instead of heavily using an operation we know ES isn't cut off for, shouldn't we just use it differently ?

**Comment by fstagni** [comment]
What do you mean with "use it differently"? Ideally we should not use it, or limit the amount of times we do it.
The JobParameters are already updated in bulk, there might be some optimizations to do but nevertheless we can't fully avoid that `update` is called. So, this ugly fix would still need to be there.

File: src/DIRAC/Core/Utilities/ElasticSearchDB.py [hunk 1]
```diff
@@ -265,9 +267,15 @@ def updateDoc(self, index: str, docID: str, body) -> dict:
         """
         sLog.debug(f"Updating document {docID} in index {index}")
         try:
-            return S_OK(self.client.update(index, docID, body))
+            self.client.update(index, docID, body)
+        except ConflictError:
+            # updates are rather "heavy" operations from ES point of view, needing seqNo to be updated.
+            # Not ideal, but we just wait and retry.
+            time.sleep(3)
+            self.client.update(index, docID, body, retry_on_conflict=3)
```
Comment (chrisburr) [inline src/DIRAC/Core/Utilities/ElasticSearchDB.py]: Is this any better than using the `retry_on_conflict` option and letting the server take care of retrying?

File: src/DIRAC/Core/Utilities/ElasticSearchDB.py [hunk 2]
```diff
@@ -265,9 +267,15 @@ def updateDoc(self, index: str, docID: str, body) -> dict:
...
         except RequestError as re:
             return S_ERROR(re)
+        return S_OK()
```
Comment (chrisburr) [inline src/DIRAC/Core/Utilities/ElasticSearchDB.py:278]: Did you intend to change the return value?

**Comment by chrisburr** [comment]
Rethinking this at a higher level, does the `Status` field need to exist in the `ElasticJobParametersDB`? Isn't it already in the `Jobs` table in the `JobDB`?

**Comment by fstagni** [comment]
> Rethinking this at a higher level, does the `Status` field need to exist in the `ElasticJobParametersDB`? Isn't it already in the `Jobs` table in the `JobDB`?

The status have been added lately to make it easier to create plots (e.g. in Grafana) based on the Status on the jobs. It is of course already in JobDB.

File: src/DIRAC/Core/Utilities/ElasticSearchDB.py (hunk 2 above)
Comment (fstagni) [reply src/DIRAC/Core/Utilities/ElasticSearchDB.py:278]: Yes, anyway it is not used.

File: src/DIRAC/Core/Utilities/ElasticSearchDB.py (hunk 1 above)
Comment (fstagni) [reply src/DIRAC/Core/Utilities/ElasticSearchDB.py]: I tried to find somewhere the information on if there's or not a delay in retrying, which I did not find. So I added it artificially (ugly, I agree).

**Comment by chaen** [comment]
I mean avoid `update`  by only inserting and taking the latest document for example.

File: src/DIRAC/Core/Utilities/ElasticSearchDB.py (hunk 1 above)
Comment (chrisburr) [reply src/DIRAC/Core/Utilities/ElasticSearchDB.py]: Is there any reason to delay? Naively I would have thought you could retry immediately.

Comment (fstagni) [reply src/DIRAC/Core/Utilities/ElasticSearchDB.py]: IIRC, and if things have not changed in the meanwhile, ES (and OpenSearch too I guess) for each update deletes the old document and then indexes a new one. The `seqNo` that is mentioned in the callstack keeps track of the version of the document. So basically an update is potentially 3 times slower than adding a new document (I am of course simplifying...). 

If we wait we have more chance that the previous update finished. I am anyway not even convinced this is needed...

**Comment by atsareg** [comment]
I think this fix is OK as it can help to solve the initial problem - to be confirmed in real operations.

**Review by atsareg — APPROVED** [review]

**Pull request merged**